remove `pytorch-{cpu,gpu}` compat outputs

h-vetinari opened the pull request:
While [reviewing](https://github.com/conda-forge/pytorch-cpu-feedstock/pull/231#discussion_r1675188481) another PR, I checked the situation with regarding `pytorch-{cpu,gpu}` again, and IMO it's way past time to remove them. Quoting from there

>  They were purely for compatibility with the old naming on the pytorch channel, but we've lost track of them completely.
> 
> In the pytorch channel, `pytorch-cpu` hasn't had a new [build](https://anaconda.org/pytorch/pytorch/files) in 5(!) years, and the gpu variant doesn't exist anymore(?!). There's [`pytorch-cuda`](https://anaconda.org/pytorch/pytorch-cuda/files), but that's only a meta-package for the various CUDA components, not pytorch itself. You can check out the upstream [recipe](https://github.com/pytorch/builder/blob/main/conda/pytorch-nightly/meta.yaml), which now also features a build variant.

We might want to wait for a new minor version to do this.

PS. We've been emitting linter hints for these since https://github.com/conda-forge/conda-forge-pinning-feedstock/commit/c00c43a2f809d3836e83cbf721bd93d58366924f

h-vetinari commented:
(CI stopped to conserve resources; we can discuss first)

hmaarrfk reviewed (approved):
But i would also just want to cancel on main.

Triggering a whole new 500MB download for end users that don't need it is silly.

hmaarrfk commented:
Lets wait for a few more stakeholders.

h-vetinari commented:
It would also be good to take another stab at fixing https://github.com/conda-forge/pytorch-cpu-feedstock/issues/155; I updated the list and opened some issues for hopefully higher visibility.

It's worth noting though that we already added linter hints for `pytorch-{cpu,gpu}` as of https://github.com/conda-forge/conda-forge-pinning-feedstock/commit/c00c43a2f809d3836e83cbf721bd93d58366924f almost a year ago

isuruf commented:
How do you propose end users select the CPU/GPU version now?

hmaarrfk commented:
> How do you propose end users select the CPU/GPU version now?

```
mamba install pytorch
```

should allow them to get the "best version" pretty consistently. 

otherwise

```
mamba install pytorch=*=cuda*
# vs
mamba install pytorch=*=cpu_*
```

isuruf commented:
```
mamba install pytorch=*=cuda*
# vs
mamba install pytorch=*=cpu_*
```

This is a bit verbose than
```
mamba install pytorch-gpu
# vs
mamba install pytorch-cpu
```
What's the harm in continuing to support these convenience outputs? It's not a burden to maintain these.


pytorch channel has the convenience outputs
```
mamba install pytorch pytorch-cuda=x
vs
mamba install pytorch cpuonly
```

h-vetinari commented:
The original intention was to match the naming of the pytorch channel so users don't accidentally coinstall the same package with different names from different channels.

So now the upstream naming has diverged, I think those outputs are barely worth it anymore.

Already, the ideal should be IMO that user choice is unnecessary, in the sense that they get the "best" implementation for their system; for the cases where a choice is necessary, the build string option is not amazing, but possible (similar to how we're advising users to set `libblas=*=*<impl>`)

If we want to continue providing convenience wrappers for that, then IMO we should match the pytorch naming (except `cpuonly`), e.g. there's also `pytorch-mutex` that they use for that. 

Matching `pytorch-cuda` would be attractive in principle, if only it didn't involve a completely different version scheme...

Long term, I'd wish for a better [API](https://github.com/conda/conda/issues/11053) to express these things, without messing with pointless outputs or build strings.

isuruf commented:
IMO, just dropping this output and wasting user's time to figure out that you now need to say `pytorch=*=*_cpu` is not worth it
especially since having these are not a huge maintenance burden.

hmaarrfk reviewed (changes requested):
I'm removing my approval since isuruf's comments reminded me of an old issue I created that still has no resolution

https://github.com/conda-forge/conda-forge.github.io/issues/1894

to my knowledge, today, there is no way to specify:

````
pytroch-gpu 2.3.1
```
should just be rewritten as
```
pytorch 2.3.1=cuda*
```
when the users types
```
mamba update --all
```

In our application, we try to identify these snags, and give users command like

```
# Do not replace the first conda command with mamba
conda uninstall --force --yes --offline old_package_name
mamba install "new_package_name>=some_lower_bound"
mamba update --all --yes
```

While I am ok with people in new environments being asked to use the new names of packages, the fact that we can't "update" old environments really bothers me since it leaves users in very strange "stuck states" of the solver.

rgommers commented:
I also think `pytorch-cpu`/`pytorch-gpu` is significantly nicer than using build strings. Removing them will be a breaking change for some users (that'd include me; xref https://github.com/conda-forge/pytorch-cpu-feedstock/issues/229#issuecomment-2493511045 for an example). And with the `pytorch` channel being deprecated now, we should just do whatever seems best for Conda-forge. 

I suggest closing this PR and keeping things as is.

h-vetinari commented:
Alright, let's close this then. My POV is still

> Long term, I'd wish for a better [API](https://github.com/conda/conda/issues/11053) to express these things, without messing with pointless outputs or build strings.

rgommers commented:
> Long term, I'd wish for a better [API](https://github.com/conda/conda/issues/11053) to express these things, without messing with pointless outputs or build strings.

No argument there from me - a nicer new syntax would be great, and if we get that then we can revisit this PR I think.

function2-llx commented:
Hello, I've been watching your discussion on removing both `pytorch-cpu` and `pytorch-gpu` in favor of `pytorch`, I appreciate your effort and would like to provide a suggestion.

Now that the official PyTorch Anaconda channel is deprecated, and users are guided to the conda-forge alternative in the [official announcement](https://github.com/pytorch/pytorch/issues/138506). However, they are explicitly suggesting `pytorch-cpu` and `pytorch-gpu` in the guidance. I'm afraid this will make more users migrate to these packages that are supposed to be deprecated some days, and significantly burden your efforts on finally removing them. In addition, there will probably be more items added to the list in https://github.com/conda-forge/pytorch-cpu-feedstock/issues/155.

Therefore, I strongly suggest that you communicate with the PyTorch team and make sure their announcement / documentation / tutorial are adjusted accordingly.

h-vetinari commented:
Thanks for the comment - it's OK though, we've decided to keep the convenience outputs anyway.
